Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+        asbd.mFormatID = m_codec;
+        asbd.mSampleRate = double(m_config.sampleRate);
+        asbd.mChannelsPerFrame = uint32_t(m_config.numberOfChannels);
+        if (m_frameDuration)
+            asbd.mFramesPerPacket = m_config.sampleRate * m_frameDuration->toDouble();
+        m_outputDescription = asbd;
+    }
+    m_codecString = codecName.isolatedCopy();
+
+    return emptyString();
+}
+
+void InternalAudioEncoderCocoa::compressedAudioOutputBufferCallback(void* object, CMBufferQueueTriggerToken)
+{
+    // We can only be called from the CoreMedia callback if we are still alive.
+    RefPtr encoder = static_cast<class InternalAudioEncoderCocoa*>(object);
+
+    InternalAudioEncoderCocoa::queueSingleton().dispatch([weakEncoder = ThreadSafeWeakPtr { *encoder }] {
+        if (auto strongEncoder = weakEncoder.get())
+            strongEncoder->processEncodedOutputs();
+    });
+}
+
+Vector<uint8_t> InternalAudioEncoderCocoa::generateDecoderDescriptionFromSample(CMSampleBufferRef sample) const
+{
+    RetainPtr formatDescription = PAL::CMSampleBufferGetFormatDescription(sample);
+    ASSERT(formatDescription);
+    const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(formatDescription.get());
+    if (!asbd)
+        return { };
+
+    if (asbd->mFormatID == kAudioFormatOpus)
+        return createOpusPrivateData(*asbd, m_converter->preSkip());
+
+    size_t cookieSize = 0;
+    auto* cookie = PAL::CMAudioFormatDescriptionGetMagicCookie(formatDescription.get(), &cookieSize);
+    if (!cookieSize)
+        return { };
+    return unsafeMakeSpan(static_cast<const uint8_t*>(cookie), cookieSize);
+}
+
+AudioEncoder::ActiveConfiguration InternalAudioEncoderCocoa::activeConfiguration(CMSampleBufferRef sample) const
+{
+    assertIsCurrent(queueSingleton());
+    ASSERT(!m_isClosed && m_converter);
+
+    RetainPtr formatDescription = PAL::CMSampleBufferGetFormatDescription(sample);
+    ASSERT(formatDescription);
+    const AudioStreamBasicDescription* const asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(formatDescription.get());
+    if (!asbd)
+        return { };
+
+    return AudioEncoder::ActiveConfiguration {
+        .codec = m_codecString,
+        .sampleRate = asbd->mSampleRate,
+        .numberOfChannels = asbd->mChannelsPerFrame,
+        .bitrate = converter()->bitRate(),
+        .description = generateDecoderDescriptionFromSample(sample)
+    };
+}
+
+void InternalAudioEncoderCocoa::processEncodedOutputs()
+{
+    assertIsCurrent(queueSingleton());
+
+    if (m_isClosed)
+        return;
+
+    while (RetainPtr cmSample = converter()->takeOutputSampleBuffer()) {
+        Ref sample = MediaSampleAVFObjC::create(cmSample.get(), 0);
+        CMBlockBufferRef rawBuffer = PAL::CMSampleBufferGetDataBuffer(cmSample.get());
+        ASSERT(rawBuffer);
+        RetainPtr buffer = rawBuffer;
+        // Make sure block buffer is contiguous.
```

Comment:
We have this pattern in a few place, it might be handy to add a utility someday.